server/authorizer/authorizer.go

```diff
@@ -97,28 +102,89 @@ func (h *Basic) ServeHTTP(w http.ResponseWriter, r *http.Request) {
 	writer.Error(w, http.StatusUnauthorized, types.NewErrorV1(types.CodeUnauthorized, types.MsgUnauthorizedError))
 }
 
-func makeInput(r *http.Request) (interface{}, error) {
+func makeInput(r *http.Request) (*http.Request, interface{}, error) {
+
 	path, err := parsePath(r.URL.Path)
 	if err != nil {
-		return nil, err
+		return r, nil, err
 	}
 
 	method := strings.ToUpper(r.Method)
 	query := r.URL.Query()
 
+	var rawBody []byte
+
+	if expectBody(r.Method, path) {
+		rawBody, err = readBody(r)
+		if err != nil {
+			return r, nil, err
+		}
+	}
+
 	input := map[string]interface{}{
 		"path":    path,
 		"method":  method,
 		"params":  query,
 		"headers": r.Header,
 	}
 
+	if len(rawBody) > 0 {
+		var body interface{}
+		if expectYAML(r) {
+			if err := util.Unmarshal(rawBody, &body); err != nil {
+				return r, nil, err
+			}
+		} else if err := util.UnmarshalJSON(rawBody, &body); err != nil {
+			return r, nil, err
+		}
+
+		// We cache the parsed body on the context so the server does not have
+		// to parse the input document twice.
+		input["body"] = body
+		ctx := SetBodyOnContext(r.Context(), body)
+		r = r.WithContext(ctx)
+	}
+
 	identity, ok := identifier.Identity(r)
 	if ok {
 		input["identity"] = identity
 	}
 
-	return input, nil
+	return r, input, nil
+}
+
+var dataAPIVersions = map[string]bool{
+	"v0": true,
+	"v1": true,
+}
+
+func expectBody(method string, path []interface{}) bool {
+	if method == http.MethodPost {
+		if len(path) == 1 {
+			s := path[0].(string)
+			return s == ""
+		} else if len(path) >= 2 {
+			s1 := path[0].(string)
+			s2 := path[1].(string)
+			return dataAPIVersions[s1] && s2 == "data"
+		}
+	}
+	return false
+}
+
+func expectYAML(r *http.Request) bool {
+	return strings.Contains(r.Header.Get("Content-Type"), "yaml")
```

Comment:
At least in our docs, the YAML criteria are stricter -- it would be `application/x-yaml`. But this is in line with what happens in server.go, so OK.

Reply:
Added a comment.